### PR TITLE
Add test for grad input and weights against keras layers

### PR DESCRIPTION
**Pull request by hkvision**
## What changes were proposed in this pull request?

Compare gradient inputs for each layer with keras
(Still has problem with batchnormalization and embedding)

## How was this patch tested?

Jenkins

### CODE REVIEW

File: pyspark/test/bigdl/test_utils.py [hunk 1]
```diff
@@ -214,6 +214,31 @@ def modelTest(self,
...
+        feed_dict = {K.learning_phase(): 1}
+        if isinstance(input_data, list):
+            for i in range(0, len(input_data)):
+                feed_dict[keras_model.input[i]] = input_data[i]
+        else:
+            feed_dict[keras_model.input] = input_data
+        keras_grad_input = sess.run(K.gradients(keras_model.output * keras_output, keras_model.input),
+                                    feed_dict=feed_dict)
+        bigdl_model.training(is_training=True)
+        bigdl_grad_input = bigdl_model.backward(input_data, bigdl_output2)
```
Comment (zhichao-li) [inline pyspark/test/bigdl/test_utils.py]: we should also check grad_weights.

Comment (hkvision) [reply pyspark/test/bigdl/test_utils.py]: Added

**Comment by hkvision** [comment]
@zhichao-li have a look at the grad input of batchnormalization and embedding?

File: pyspark/test/bigdl/keras/test_layer.py [hunk 2]
```diff
@@ -135,22 +170,22 @@ def test_atrousconvolution2d(self):
...
         layer2 = Deconvolution2D(3, 3, 3, output_shape=(None, 3, 25, 25),
                                  border_mode="valid", subsample=(2, 2),
                                  dim_ordering="th", input_shape=(3, 12, 12))
-        self.modelTestSingleLayer(input_data, layer2, dump_weights=True)
+        self.modelTestSingleLayer(input_data, layer2, dump_weights=True, rtol=1e-5, atol=1e-5)
 
     def test_maxpooling3d(self):
         input_data = np.random.random_sample([1, 3, 20, 15, 35])
-        layer = MaxPooling3D(pool_size=(2, 2, 4), strides=(3, 1, 5), dim_ordering="th",
+        layer = MaxPooling3D(pool_size=(2, 2, 3), strides=(1, 2, 3), dim_ordering="th",
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras/test_layer.py]: why change this?

File: pyspark/test/bigdl/keras/test_layer.py [hunk 3]
```diff
@@ -191,16 +226,16 @@ def test_globalmaxpooling1d(self):
...
         input_data = np.random.random_sample([2, 6, 20, 15, 35])
-        layer = AveragePooling3D(pool_size=(2, 3, 4), strides=(3, 1, 5), dim_ordering="th",
+        layer = AveragePooling3D(pool_size=(2, 3, 4), strides=(1, 2, 3), dim_ordering="th",
                                  border_mode="valid", input_shape=(3, 20, 15, 35))
         self.modelTestSingleLayer(input_data, layer)
 
     def test_averagepooling2d(self):
         input_data = np.random.random_sample([1, 3, 20, 20])
-        layer = lambda: AveragePooling2D(pool_size=[2, 3], strides=[4, 2],
+        layer = lambda: AveragePooling2D(pool_size=[2, 3], strides=[1, 2],
```
Comment (zhichao-li) [inline pyspark/test/bigdl/keras/test_layer.py]: add a new test instead of change it if needed?

File: pyspark/test/bigdl/test_utils.py [hunk 4]
```diff
@@ -261,6 +302,7 @@ def modelTestSingleLayer(self,
                 functional_api=api,
                 random_weights=random_weights,
                 dump_weights=dump_weights,
+                weight_converter=weight_converter,
```
Comment (zhichao-li) [inline pyspark/test/bigdl/test_utils.py]: why we need a weight_converter here?

File: pyspark/test/bigdl/keras/test_layer.py (hunk 2 above)
Comment (hkvision) [reply pyspark/test/bigdl/keras/test_layer.py]: To compute grad input, there are constraints on pool size

File: pyspark/test/bigdl/test_utils.py [hunk 5]
```diff
@@ -214,13 +215,51 @@ def modelTest(self,
...
+        feed_dict = {K.learning_phase(): 1}
+        if isinstance(input_data, list):  # for merge layers
+            for i in range(0, len(input_data)):
+                feed_dict[keras_model.input[i]] = input_data[i]
+        else:
+            feed_dict[keras_model.input] = input_data
+        keras_grad_input = sess.run(K.gradients(keras_model.output * keras_output, keras_model.input),
+                                    feed_dict=feed_dict)
+        bigdl_grad_input = bigdl_model.backward(input_data, bigdl_output2)
+        if isinstance(bigdl_grad_input, list):  # for merge layers
```
Comment (zhichao-li) [inline pyspark/test/bigdl/test_utils.py]: Extract this into two parts. one is for getting the grad, the other is for comparison.

File: pyspark/test/bigdl/test_utils.py (hunk 4 above)
Comment (hkvision) [reply pyspark/test/bigdl/test_utils.py]: Oh, sorry. I forgot to remove it.

File: pyspark/test/bigdl/test_utils.py (hunk 5 above)
Comment (hkvision) [reply pyspark/test/bigdl/test_utils.py]: You mean we use two functions separately for doing this? But when comparing the forward output, we are not doing in two parts.
